### PR TITLE
feat: Item Storage GUI

### PR DESCRIPTION
**Pull request by ShadowCat117**
Store encoded items in a personal vault.
- Left click item to go to sharing GUI.
- Shift left click to remove from vault.
- Right click to change to a different category.
- Add categories via button and input
- Delete categories via button, click to move items from category to uncategorised, shift left click to delete items too
- Scrolling when mouse is above the first row of items will scroll through categories, if below will scroll items

https://github.com/Wynntils/Artemis/assets/8337467/10366436-5b7e-4e74-8687-85a481387a98

### CODE REVIEW

File: common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java [hunk 1]
```diff
@@ -0,0 +1,528 @@
...
+                            this.topPos + 10 + 8,
+                            99,
+                            CommonColors.WHITE,
+                            HorizontalAlignment.CENTER,
+                            VerticalAlignment.MIDDLE,
+                            TextShadow.NONE);
+        }
+
+        // Draw a box around the selected item to let the user know which item they are moving
+        if (selectedItemSlot != null && currentCategory.equals(originalCategory)) {
```
Comment (ShadowCat117) [inline common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java]: Realising now this won't work, if you scroll at all it'll highlight wrong item

File: common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java [hunk 2]
```diff
@@ -0,0 +1,528 @@
...
+        // Left+Shift deletes the item from storage
+        // Right click selects it for category change
+        if (mouseButton == 0) {
+            if (KeyboardUtils.isShiftDown()) {
+                deleteItem(encodedItems.get(slot.index));
+            } else {
+                Optional<WynnItem> wynnItemOpt = Models.Item.getWynnItem(slot.getItem());
+
+                if (wynnItemOpt.isPresent()) {
+                    McUtils.mc().setScreen(ItemSharingScreen.create(wynnItemOpt.get(), slot.getItem(), true));
```
Comment (ShadowCat117) [inline common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java:185]: Left click action probably not needed, can just use F3 on item and ChatItemFeature can handle it

**Comment by ShadowCat117** [comment]
Also need a way to edit category names and potentially the ability to select more than 1 item at a time

**Comment by kristofbolyai** [comment]
> Also need a way to edit category names and potentially the ability to select more than 1 item at a time

Perhaps the container search bar could be added as well.

File: common/src/main/java/com/wynntils/screens/itemsharing/ItemSharingScreen.java [hunk 3]
```diff
@@ -162,17 +184,67 @@ private void refreshPreview() {
...
+                    }
+                }
+
+                ItemStack itemStackToSave = itemStack;
+
+                // Gear items can have their item changed by cosmetics so we need to get their original item
+                // FIXME: Does not work for crafted gear
+                if (wynnItem instanceof GearItem gearItem) {
+                    itemStackToSave = new FakeItemStack(gearItem, "From " + McUtils.playerName() + "'s vault");
+                }
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/screens/itemsharing/ItemSharingScreen.java]: So what's the problem with crafted gear/consumables?

File: common/src/main/java/com/wynntils/screens/itemsharing/ItemSharingScreen.java [hunk 4]
```diff
@@ -162,17 +184,67 @@ private void refreshPreview() {
...
+                        : -1;
+
+                // Regular ItemStack can't be converted to json so store the tags needed
+                // to recreate it
+                SavedItemStack savedItemStack = new SavedItemStack(
+                        Item.getId(itemStackToSave.getItem()),
+                        itemStackToSave.getTag().getInt("Damage"),
+                        itemStackToSave.getTag().getInt("HideFlags"),
+                        itemStackToSave.getTag().getBoolean("Unbreakable"),
+                        color);
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/screens/itemsharing/ItemSharingScreen.java]: If you want to save non-chat encodable items, you would also need to save their lore. Not sure if this is an useful case.

File: common/src/main/java/com/wynntils/screens/itemsharing/ItemSharingScreen.java [hunk 5]
```diff
@@ -235,11 +307,25 @@ private void addSharingOptions() {
                 .size(shareButtonWidth, 20)
                 .build()));
 
-        options.add(this.addRenderableWidget(new Button.Builder(
-                        Component.translatable("screens.wynntils.itemSharing.save"), (b) -> shareItem("save"))
+        // If an item has already been saved then this button will act as easy access to their saved items
+        Component saveButtonMessage = savedItem
+                ? Component.translatable("screens.wynntils.itemSharing.openVault")
+                : Component.translatable("screens.wynntils.itemSharing.save");
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/screens/itemsharing/ItemSharingScreen.java:276]: Maybe "Save to Vault"?

File: common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java [hunk 6]
```diff
@@ -0,0 +1,528 @@
...
+import net.minecraft.world.inventory.Slot;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.item.ItemStack;
+import org.lwjgl.glfw.GLFW;
+
+public final class SavedItemsScreen extends WynntilsContainerScreen<SavedItemsMenu> {
+    private static final int ITEMS_PER_ROW = 7;
+    private static final int MAX_ITEMS = 49;
+    private static final int SCROLL_AREA_HEIGHT = 128;
+    private static final String DEFAULT_CATEGORY = "Uncategorized";
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java]: Probably better referenced from the feature. Or even better. Move the whole storage of items to `ItemVaultService`, and store that constant there, along with the items. You could also make convenience methods in the service, without accessing the feature.

File: common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java (hunk 2 above)
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java:185]: Hm. I think I probably like the left click option as well, if we have no other use for left click.

File: common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java [hunk 7]
```diff
@@ -0,0 +1,528 @@
...
+            draggingScroll = true;
+            return true;
+        }
+
+        return super.mouseClicked(mouseX, mouseY, button);
+    }
+
+    @Override
+    public boolean mouseDragged(double mouseX, double mouseY, int button, double dragX, double dragY) {
+        if (!draggingScroll) return false;
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java]: Shouldn't you return super here as well?

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java]: (usually it doesn't matter, but theoretically that would be the correct code)

File: common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java [hunk 8]
```diff
@@ -0,0 +1,528 @@
...
+                this.leftPos + 36, this.topPos + 5, 99, 16, null, (TextboxScreen) McUtils.mc().screen, categoryInput);
+
+        this.addRenderableWidget(categoryInput);
+    }
+
+    private void addCategory() {
+        if (addingCategory) {
+            if (!categoryInput.getTextBoxInput().isEmpty()) {
+                // Save new category
+                Models.ItemEncoding.savedItems.get().put(categoryInput.getTextBoxInput(), new HashMap<>());
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java]:
```suggestion
                Models.ItemEncoding.savedItems.get().put(categoryInput.getTextBoxInput(), new TreeMap<>());
```

File: common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java (hunk 6 above)
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java]: Hm. I see now that it is stored in the encoding model, not the feature. Anyways, let's move away from that, and move to the service.

**Comment by kristofbolyai** [comment]
Let's get to some UI/UX improvements I miss:
- When I create a new category, it should automatically switch to it, instead of having to manually do that.
- The help button might be better put to the screen itself, probably the left corner at the bottom of the screen, using the textures of the other buttons. Currently it is very hard to notice.


Seems like part of the conversation got lost when we talked about details in discord. I think it would be really useful to allow items to be in multiple categories. Config wise, it should not be hard, although you might as well save the categories in the SavedItem object, and just construct a map from the list itself. You can also save the `base64` of the item there as well, and construct a map later. You could even add a custom deserializer class for it. I could do that, after I finish the build system changes.

Also just want to state that if we can't finish this today, we will delay the releases. No hurry, we want this to work nicely.

**Comment by magicus** [comment]
Is this some kind of list of contents in your bank? That is, the vault does not contain any actual stuff, but just description of items?

I'm not fully sure I understand the purpose. (I realize I could go into discord and search, but if you can summarize it in a sentence or two, I think that would be good for the users as well)

Can it be automatically populated from the bank?

Or is it a track record of all items you have ever owned? If so, can it be automatically populated?

Or is it kind of a side-loaded selling system, where you can offer trades for certain of your items?

**Comment by magicus** [comment]
In any case, I think "Vault" is a somewhat misleading name. It sounds like a safe storage, but in fact you are not storing anything at all. It would be sad indeed if a user mistook this for a safe solution to store their real item, and somehow got the real item lost.

File: common/src/main/java/com/wynntils/screens/itemsharing/ItemSharingScreen.java (hunk 3 above)
Comment (ShadowCat117) [reply common/src/main/java/com/wynntils/screens/itemsharing/ItemSharingScreen.java]: Crafted armour/weapons still show their cosmetics/the shield icon if hidden via silverbull. Everything else works fine

**Comment by ShadowCat117** [comment]
> When I create a new category, it should automatically switch to it, instead of having to manually do that.

I did consider that, but since it'll just be empty and then you'll have to go back to where your items was to select for movement I decided against it, perhaps if you have any selected items it will automatically move them to the new category on creation?

**Comment by kristofbolyai** [comment]
> > When I create a new category, it should automatically switch to it, instead of having to manually do that.
> 
> 
> 
> I did consider that, but since it'll just be empty and then you'll have to go back to where your items was to select for movement I decided against it, perhaps if you have any selected items it will automatically move them to the new category on creation?
> 
> 

Yes, sure!

**Comment by ShadowCat117** [comment]
> Is this some kind of list of contents in your bank? That is, the vault does not contain any actual stuff, but just description of items?
> 
> I'm not fully sure I understand the purpose. (I realize I could go into discord and search, but if you can summarize it in a sentence or two, I think that would be good for the users as well)

It's essentially a way to save any of your items for easier access to share them, so say you store all your mythics in the vault, whenever you want to make a shout to put them up for sale you can just access the vault and copy it from there, without having to go back to your bank.

File: common/src/main/java/com/wynntils/screens/itemsharing/ItemSharingScreen.java (hunk 3 above)
Comment (ShadowCat117) [reply common/src/main/java/com/wynntils/screens/itemsharing/ItemSharingScreen.java]: Actually, it probably also doesn't work for accessories and consumables as for some reason Wynn changes non armour to the shield item if they're in certain container slots (usually the bottom right corner)
image

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/screens/itemsharing/ItemSharingScreen.java]: That's acceptable for now.

**Comment by ShadowCat117** [comment]
Haven't tested the final commit much as I was rushing a bit to try and get it done today so I'll do that now, even if it looks fine don't merge it just yet. (I also forgot I left a todo)

File: common/src/main/java/com/wynntils/services/itemvault/ItemVaultService.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+import com.wynntils.services.itemvault.type.SavedItem;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+public class ItemVaultService extends Service {
+    private static final String DEFAULT_CATEGORY = "Uncategorized";
+
+    @Persisted
+    public final Storage<Set<SavedItem>> savedItems = new Storage<>(new HashSet<>());
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/services/itemvault/ItemVaultService.java]:
```suggestion
    public final Storage<Set<SavedItem>> savedItems = new Storage<>(new TreeSet<>());
```

File: common/src/main/java/com/wynntils/services/itemvault/ItemVaultService.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+    @Persisted
+    public final Storage<Set<String>> categories = new Storage<>(new HashSet<>(List.of(DEFAULT_CATEGORY)));
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/services/itemvault/ItemVaultService.java]:
```suggestion
    public final Storage<Set<String>> categories = new Storage<>(new TreeSet<>(List.of(DEFAULT_CATEGORY)));
```

File: common/src/main/java/com/wynntils/services/itemvault/type/SavedItem.java [hunk 11]
```diff
@@ -0,0 +1,16 @@
+/*
+ * Copyright © Wynntils 2024.
+ * This file is released under LGPLv3. See LICENSE for full license details.
+ */
+package com.wynntils.services.itemvault.type;
+
+import java.util.Set;
+
+public record SavedItem(
+        String base64, Set<String> categories, int itemID, int damage, int hideFlags, boolean unbreakable, int color)
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/services/itemvault/type/SavedItem.java]: Perhaps you could make an inner class here, that stores all `ItemStack` related  info.

File: common/src/main/java/com/wynntils/screens/itemsharing/widgets/SavedItemsHelpButton.java [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+import com.wynntils.screens.base.widgets.WynntilsButton;
+import com.wynntils.utils.mc.ComponentUtils;
+import com.wynntils.utils.mc.McUtils;
+import com.wynntils.utils.render.RenderUtils;
+import com.wynntils.utils.render.Texture;
+import java.util.List;
+import net.minecraft.client.gui.GuiGraphics;
+import net.minecraft.network.chat.Component;
+
+public class SavedItemsHelpButton extends WynntilsButton {
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/screens/itemsharing/widgets/SavedItemsHelpButton.java]: This class could extend `SavedItemsButton`?

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/screens/itemsharing/widgets/SavedItemsHelpButton.java]: Or the other way around, either way, I see it as almost 1 to 1 code duplication.

**Review by kristofbolyai — COMMENTED** [review]
Quick review, I'll be back. No rush.

File: common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java [hunk 13]
```diff
@@ -0,0 +1,607 @@
...
+        super.doInit();
+        this.leftPos = (this.width - Texture.VAULT.width()) / 2;
+        this.topPos = (this.height - Texture.VAULT.height()) / 2;
+
+        // region Side buttons
+        this.addRenderableWidget(new SavedItemsButton(
+                this.leftPos + 8,
+                this.topPos + 22,
+                (b) -> {
+                    if (b == 0) {
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java]: Use the GLFW constant for the mosue button here.

File: common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java [hunk 14]
```diff
@@ -0,0 +1,607 @@
...
+                        Component.translatable("screens.wynntils.savedItems.deleteTooltip2")),
+                Texture.VAULT_DELETE));
+
+        this.addRenderableWidget(new SavedItemsButton(
+                this.leftPos + 8,
+                this.topPos + 50,
+                (b) -> {
+                    if (b == 0) {
+                        moveSelectedItems();
+                    } else if (b == 1) {
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java]: Same here

File: common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java [hunk 15]
```diff
@@ -0,0 +1,607 @@
...
+    @Override
+    protected void slotClicked(Slot slot, int slotId, int mouseButton, ClickType type) {
+        if (slot == null) return;
+        if (slot.getItem() == ItemStack.EMPTY) return;
+
+        // Left click goes to sharing menu
+        // Left+Shift deletes the item from storage
+        // Right click selects it for category change
+        // Right+Shift will also deletes it from the current category when moved
+        if (mouseButton == 0) {
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java]: and here

File: common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java [hunk 16]
```diff
@@ -0,0 +1,607 @@
...
+            } else {
+                Optional<WynnItem> wynnItemOpt = Models.Item.getWynnItem(slot.getItem());
+
+                if (wynnItemOpt.isPresent()) {
+                    McUtils.mc().setScreen(ItemSharingScreen.create(wynnItemOpt.get(), slot.getItem(), true));
+                } else {
+                    McUtils.sendMessageToClient(Component.translatable("screens.wynntils.savedItems.unableToShare"));
+                }
+            }
+        } else if (mouseButton == 1) {
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java]: and here

File: common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java [hunk 17]
```diff
@@ -0,0 +1,607 @@
...
+            Services.ItemVault.categories.touched();
+
+            if (!selectedItems.isEmpty()) {
+                currentCategory = newCategory;
+                moveSelectedItems();
+            }
+        } else if (editingCategory) {
+            // Add renamed category and remove previous name
+            Services.ItemVault.categories.get().add(newCategory);
+            Services.ItemVault.categories.get().remove(currentCategory);
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java]: This does not preserve the category order, maybe we should do that?

File: common/src/main/resources/assets/wynntils/lang/en_us.json [hunk 18]
```diff
@@ -1932,6 +1935,21 @@
   "screens.wynntils.poiManagementGui.setMarkers": "Set Markers",
   "screens.wynntils.poiManagementGui.setMarkers.tooltip": "Sets a marker at all selected POIs",
   "screens.wynntils.poiManagementGui.undo": "Undo delete",
+  "screens.wynntils.savedItems.addTooltip1": "Click to add a new category to your vault.",
+  "screens.wynntils.savedItems.addTooltip2": "Enter the name of the category you want to add then hit enter or click this button again to save it.",
+  "screens.wynntils.savedItems.changeCategoryTooltip1": "Left click to move the selected item to the current category.",
+  "screens.wynntils.savedItems.changeCategoryTooltip2": "Right click to cancel.",
+  "screens.wynntils.savedItems.deleteTooltip1": "Left click to delete current category and move all items in the category to no category",
+  "screens.wynntils.savedItems.deleteTooltip2": "Shift-Left click to delete current category and delete all items in current category",
```
Comment (kristofbolyai) [inline common/src/main/resources/assets/wynntils/lang/en_us.json]:
```suggestion
  "screens.wynntils.savedItems.deleteTooltip1": "Left click to delete current category and move all items in the category to no category.",
  "screens.wynntils.savedItems.deleteTooltip2": "Shift-Left click to delete current category and delete all items in current category.",
```

More consistent this way

File: common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java [hunk 19]
```diff
@@ -0,0 +1,607 @@
...
+        } else if (!currentCategory.equals(Services.ItemVault.getDefaultCategory())) {
+            // Remove category from all items and add default
+            for (SavedItem savedItem : savedItems) {
+                savedItem.categories().remove(currentCategory);
+                savedItem.categories().add(Services.ItemVault.getDefaultCategory());
+            }
+
+            Services.ItemVault.savedItems.store(savedItems);
+            Services.ItemVault.savedItems.touched();
+        }
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java]: These two could be merged if the if is moved inside, right?

File: common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java [hunk 20]
```diff
@@ -0,0 +1,607 @@
...
+    private void populateItems() {
+        List<ItemStack> items = new ArrayList<>();
+        List<ItemStack> selected = new ArrayList<>();
+        encodedItems = new ArrayList<>();
+        selectedSlots = new ArrayList<>();
+
+        // Clear current items
+        for (int i = 0; i < MAX_ITEMS; i++) {
+            this.menu.setItem(i, 0, ItemStack.EMPTY);
+        }
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java]: Perhaps implement a this.menu.clear method, if it does not exist yet.

File: common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java [hunk 21]
```diff
@@ -0,0 +1,607 @@
...
+
+        for (int i = rowOffset; i < (MAX_ITEMS + rowOffset); i++) {
+            if (i >= savedItems.size()) break;
+
+            SavedItem savedItem = savedItems.get(i);
+            encodedItems.add(savedItem);
+
+            // Create the ItemStack from the store tag values
+            ItemStack itemStack;
+            itemStack = new ItemStack(Item.byId(savedItem.itemStackInfo().itemID()), 1);
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java]: could be just one line

File: common/src/main/java/com/wynntils/services/itemvault/ItemVaultService.java [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+import java.util.TreeSet;
+
+public class ItemVaultService extends Service {
+    private static final String DEFAULT_CATEGORY = "Uncategorized";
+
+    @Persisted
+    public final Storage<Set<SavedItem>> savedItems = new Storage<>(new TreeSet<>());
+
+    @Persisted
+    public final Storage<Set<String>> categories = new Storage<>(new TreeSet<>(List.of(DEFAULT_CATEGORY)));
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/services/itemvault/ItemVaultService.java]: I think you could move a lot of logic away from the screen by creating helper methods for saving/loading/updating the items. Like `addItem`, `setCategory`, `addCategory`, etc.

File: common/src/main/java/com/wynntils/services/itemvault/type/SavedItem.java [hunk 23]
```diff
@@ -0,0 +1,17 @@
...
+package com.wynntils.services.itemvault.type;
+
+import java.util.Set;
+
+public record SavedItem(String base64, Set<String> categories, ItemStackInfo itemStackInfo)
+        implements Comparable<SavedItem> {
+    @Override
+    public int compareTo(SavedItem other) {
+        return this.base64.compareTo(other.base64);
+    }
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/services/itemvault/type/SavedItem.java]: Maybe compare the item stack info as well? Or even the categories. I don't think that would hurt.

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/services/itemvault/type/SavedItem.java]: You can use `ComparisonChain`.

**Review by kristofbolyai — COMMENTED** [review]
You have made really good progress here @JamieCallan117! But in no way do I want to rush you, or this PR. How do you feel about delaying the release until tomorrow evening, at least? That gives us one more day to (self-)review, think, and fix up everything.

As for this PR, it's looking great. I have an idea on making the service/saved item data a bit more accessible, so I'll be working on implementing that here, tomorrow.

**Comment by ShadowCat117** [comment]
> How do you feel about delaying the release until tomorrow evening, at least? 

If you're alright with that I think it'd be for the best, I could implement all your suggested reviews tonight but I'd like to test it a bit more for potential issues

File: common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java (hunk 17 above)
Comment (ShadowCat117) [reply common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java]: Currently they're just ordered A-z. Do we want to be able to customise the order, could make right clicking the navigaton buttons shift their position or something

**Comment by kristofbolyai** [comment]
Seems like this tooltip is outdated, and does not include shift-to-move.
image

**Comment by kristofbolyai** [comment]
@JamieCallan117 I've pushed [refactor: custom json serialization for SavedItem](https://github.com/Wynntils/Artemis/pull/2312/commits/594a2611888080f4bf90e147ceea885f0dd85ead) for you, I'll leave the other stuff to you.

File: common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java (hunk 17 above)
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java]: Hm, okay, that should work.

File: common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java (hunk 19 above)
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/screens/itemsharing/SavedItemsScreen.java]: Hm, let's not do that. It's messy.

File: common/src/main/java/com/wynntils/services/itemvault/type/SavedItem.java (hunk 23 above)
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/services/itemvault/type/SavedItem.java]: Hm. Maybe it's not need afterall.

**Comment by kristofbolyai** [comment]
I've went ahead and pushed some of my review requests, so it's easier for you.

**Comment by kristofbolyai** [comment]
Can we make sure we add a separate keybind for adding items to the vault? One of us should commit that, but I don't have time for it at the moment.

**Comment by magicus** [comment]
> It's essentially a way to save any of your items for easier access to share them, so say you store all your mythics in the vault, whenever you want to make a shout to put them up for sale you can just access the vault and copy it from there, without having to go back to your bank.

So maybe it's an Item Catalogue or Item Register instead of a "vault"?

**Comment by kristofbolyai** [comment]
> > It's essentially a way to save any of your items for easier access to share them, so say you store all your mythics in the vault, whenever you want to make a shout to put them up for sale you can just access the vault and copy it from there, without having to go back to your bank.
> 
> So maybe it's an Item Catalogue or Item Register instead of a "vault"?

Perhaps. The vault name seemed really "fitting", but I can see your concern about it messing with some people's idea of what it does. Let me think about what name to use.

**Comment by magicus** [comment]
To me, "vault" signals "safe storage". So putting something in the vault seems like safeguarding it, when this is in fact just a record of "favorite" items. I worry that it is misleading, and not helpful.

Maybe just "favorite" then? Bookmark? Register? Record? Catalogue? Directory?

**Comment by kristofbolyai** [comment]
Okay. Let's rename to **Item Record** then!

File: common/src/main/resources/assets/wynntils/textures/item_storage/vault.png
Comment (ShadowCat117) [inline common/src/main/resources/assets/wynntils/textures/item_storage/vault.png:1]: You don't seem to have relocated the textures to the new directory? They've just been deleted

File: common/src/main/java/com/wynntils/features/chat/ChatItemFeature.java [hunk 24]
```diff
@@ -169,10 +179,25 @@ private void onInventoryPress(Slot hoveredSlot) {
...
+                makeChatPrompt(wynnItemOpt.get());
+            }
         } else {
-            makeChatPrompt(wynnItemOpt.get());
+            ItemStack itemStackToSave = hoveredSlot.getItem();
+
+            // Gear items can have their item changed by cosmetics so we need to get their original item
+            // FIXME: Does not work for crafted gear
+            if (wynnItemOpt.get() instanceof GearItem gearItem) {
+                itemStackToSave = new FakeItemStack(gearItem, "From " + McUtils.playerName() + "'s vault");
```
Comment (ShadowCat117) [inline common/src/main/java/com/wynntils/features/chat/ChatItemFeature.java]: Few more references to vault to be changed

File: common/src/main/java/com/wynntils/features/chat/ChatItemFeature.java [hunk 25]
```diff
@@ -169,10 +179,25 @@ private void onInventoryPress(Slot hoveredSlot) {
...
+            }
+
+            // Item name is passed in since it is lost in the instanceof check above and looks nicer
+            // saying "Saved Gale's Force to your vault" than "Saved Bow to your vault"
```
Comment (ShadowCat117) [inline common/src/main/java/com/wynntils/features/chat/ChatItemFeature.java]: Here

File: common/src/main/java/com/wynntils/screens/itemsharing/ItemSharingScreen.java [hunk 26]
```diff
@@ -162,17 +182,32 @@ private void refreshPreview() {
...
+        switch (target) {
+            case "guild" -> Handlers.Command.sendCommand("g " + encodedItem.toUtf16String());
+            case "party" -> Handlers.Command.sendCommand("p " + encodedItem.toUtf16String());
+            case "save" -> {
+                ItemStack itemStackToSave = itemStack;
+
+                // Gear items can have their item changed by cosmetics so we need to get their original item
+                // FIXME: Does not work for crafted gear
+                if (wynnItem instanceof GearItem gearItem) {
+                    itemStackToSave = new FakeItemStack(gearItem, "From " + McUtils.playerName() + "'s vault");
```
Comment (ShadowCat117) [inline common/src/main/java/com/wynntils/screens/itemsharing/ItemSharingScreen.java]: And here

**Comment by kristofbolyai** [comment]
Should be fixed @JamieCallan117. Anything else?

**Comment by ShadowCat117** [comment]
I think it looks good now!

**Pull request merged**